### ISSUE TITLE
Provide Beans for JSON Response from Relay Webhooks

### ISSUE THREAD

**Issue by jonturgeon** [SparkPost/java-sparkpost#62]
Would be nice to have a set of POJO's that could be used with GSON for decoding the JSON Relay Webhook response.

**Comment by yepher** [comment]
It is a good idea and I am happy to create the POJO's

I assume the user of the POJO's would have the responsibility to run their own HTTP server and instantiate the proper Event class for deserialization?

**Comment by yepher** [comment]
I have a couple of generators that maybe easily adapted to be able to create the POJO's for the Webhook events. They are:

* [Webhook Documentor](https://github.com/SparkPost/momentum_sample_policy/tree/master/webhook-document-generator)
* an ObjC generator that is not checked in yet

It might be pretty easy to adapt those to generate at least the initial Webhook POJOs

**Comment by jonturgeon** [comment]
Hi Chris, yes to both your questions. My webapp both needs to send and receive emails. The email content is sort of structured data (it is a position reporting system for sailing yachts around the world). I use GSON for other JSON data requests and since you guys already have a Java lib I thought it might be nice to have the POJO's to make getting the relay webhook data in a Java friendly manner.


________________________________

**Comment by jonturgeon** [comment]
Hi again, I used this site to get my initial POJO's http://pojo.sodhanalibrary.com/Convert


I fed it the sample response from the relay webhook documentation. Just thought it might be nicer to have a set of POJO's that came from SparkPost to make sure we are not all missing something.


________________________________

**Comment by yepher** [comment]
yes, I totally agree with you. I will be nice to have those directly in the SparkPost Library. My comments above are really more about can I do something to help me generate code on other projects I need to do at the same time as I get this implemented in the Java library.

**Comment by yepher** [comment]
I should also note I gladly welcome Pull Requests if this is something you want to take on.

**Comment by jonturgeon** [comment]
Hi again, although I have been whacking away at Java for years, I am completely new to SparkPost and I would not want to put something out with that little bit of knowledge. If you want I can send you my beans so far, they seem to be working for my single test case but they are by no means finished. I also do this development as a bit of a hobby, I left my corporate job 6 years ago to sail around the world, I am doing this website as sort of a way to keep my head in it. I would also be willing to incorporate any beans you create and drop mine and see how they work for me.


________________________________

**Comment by yepher** [comment]
I guess I had better meet you on your Yacht so we can pair program it then ;)

Kidding aside, I would be glad to look at what you have. I will work on adding Webhook events as a first class citizen to the library as you suggested.

**Comment by jonturgeon** [comment]
Attached are my work in progress beans and below is the code I use to populate them


        //Stripes puts the JSON response in the parameter map
        String relayJSON = this.getContext().getRequest().getParameterMap().entrySet().iterator().next().toString();

        //The JSON is a bit wonky, need to be "lenient"
        JsonReader reader = new JsonReader(new StringReader(relayJSON));
        reader.setLenient(true);
        RelayWebhook[] relayWebhooks = new Gson().fromJson(reader, RelayWebhook[].class);



________________________________

**Comment by yepher** [comment]
I've created a branch with the initial pass at creating webhook events [here](https://github.com/SparkPost/java-sparkpost/tree/feature/issue_62_webhook_events/libs/sparkpost-lib/src/main/java/com/sparkpost/model/webhook/event)

Still need to test them to see if they work.
